### PR TITLE
#3500 issue, added optional dtype arg

### PR DESCRIPTION
**Pull request by tanmaymunjal**
Fixes #   Add optional dtype argument to avoid forced reduction to np.float32(Issue #3500)

Changes made in this Pull Request:
 - 
Added optional dtype arg to check_box() func 

PR Checklist
------------
 - [Did it ] Tests?
 - [Updated ] Docs?
 - [Yes] CHANGELOG updated?
 - [ #3500 ] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/lib/util.py [hunk 1]
```diff
@@ -2412,13 +2412,15 @@ def check_box(box):
         triclinic and must be provided in the same format as returned by
         :attr:`MDAnalysis.coordinates.timestep.Timestep.dimensions`:
         ``[lx, ly, lz, alpha, beta, gamma]``.
+    dtype: numpy.dtype
+        The data type of the inputted and returned box matrix.
```
Comment (tylerjereddy) [inline package/MDAnalysis/lib/util.py]: There's a typo here (`inputted`), but more broadly I'm not so sure we should impose any restriction on the input `dtype`. For example, someone may want to downcast, and I'm not sure we should stop them--I don't think the code currently checks for that? Maybe just documenting `dtype` similarly to `triclinic_vectors` would make sense?

**Review by tylerjereddy — COMMENTED** [review]
Do you think it would make sense to add a regression test for respecting the return type specified? For example, this picks up the issue for one of the two functions:

```diff
--- a/testsuite/MDAnalysisTests/lib/test_util.py
+++ b/testsuite/MDAnalysisTests/lib/test_util.py
@@ -326,16 +326,17 @@ class TestMatrixOperations(object):
             box = np.zeros(6, dtype=np.float32)
         return box
 
+    @pytest.mark.parametrize('dtype', [np.float32, np.float64])
     @pytest.mark.parametrize('lengths', comb_wr([-1, 0, 1, 2], 3))
     @pytest.mark.parametrize('angles',
                              comb_wr([-10, 0, 20, 70, 90, 120, 180], 3))
-    def test_triclinic_vectors(self, lengths, angles):
-        box = lengths + angles
+    def test_triclinic_vectors(self, lengths, angles, dtype):
+        box = np.asarray(lengths + angles, dtype=dtype)
         ref = self.ref_trivecs(box)
         res = mdamath.triclinic_vectors(box)
         assert_array_equal(res, ref)
-        # check for default dtype:
-        assert res.dtype == np.float32
+        # check for dtype pass-through:
+        assert res.dtype == dtype
         # belts and braces, make sure upper triangle is always zero:
         assert not(res[0, 1] or res[0, 2] or res[1, 2])

```

That said, the number of parameters and test time do increase a fair bit for me locally with that, so perhaps Richard might prefer a smaller/simpler test to deal with the type preservation separately?

**Comment by hmacdope** [comment]
We also have another open PR with this fix #3544 and ongoing discussion in   #3707 #3545. Basically we are waiting on further discussion as to whether we allow np.float64 coordinates etc.

**Comment by tylerjereddy** [comment]
Good point, I suggest we close this then.

It looks like for this specific set of fairly minor changes Richard and I are both already in favor--I can't see many problems arising from these changes in isolation, even if we didn't go all in on custom dtypes everywhere. It sounds like there was also demand for multi-precision at a recent conference--could that context be added to one of the discussion issues? Probably to gh-3707?

File: package/MDAnalysis/lib/util.py (hunk 1 above)
Comment (tanmaymunjal) [reply package/MDAnalysis/lib/util.py]: 1) Probably makes sense to have some tests to it, your regression test changes seem great to me

2) Will fix the typo 

3) Don't really think there are any restrictions on downcasting dtype from np.float32 that I added in

Comment (tanmaymunjal) [reply package/MDAnalysis/lib/util.py]: Made a few changes in line with the points

**Comment by hmacdope** [comment]
Hi @tanmaymunjal. This PR is very similar to an existing PR with this fix #3544. We will aim to merge that first and see if there is anything outstanding that you have picked up here. Thanks for your understanding.

**Comment by tanmaymunjal** [comment]
Sure, @hmacdope , np. I have also been working on #2046, lib.distances.distance_array() issue, almost done. Any idea if there is any current PR or discussion around it or should I send a PR for it?

**Comment by hmacdope** [comment]
> Sure, @hmacdope , np. I have also been working on #2046, lib.distances.distance_array() issue, almost done. Any idea if there is any current PR or discussion around it or should I send a PR for it?

@tanmaymunjal which subissue on the checklist are you looking to address?

**Comment by tanmaymunjal** [comment]
@hmacdope, the lib,distances.distance_array() issue and maybe the false sharing one as well,15th and 16th issue I think under the code issues headline

**Comment by hmacdope** [comment]
The distance array issue no one is working on as far as I know. For openMP some changes were made recently by @RMeli  in #3706 and #3728, so some of that may have been addressed but I am not certain.

**Comment by tanmaymunjal** [comment]
Ok, Thanks!

**Comment by orbeckst** [comment]
@hmacdope is this PR still relevant or should we close it?

**Comment by hmacdope** [comment]
> @hmacdope is this PR still relevant or should we close it?

Closeable

**Pull request closed without merging**